query-engine/query-engine-node-api/src/logger.rs

```diff
@@ -1,99 +1,85 @@
-
-        Self { subscriber }
+        JsonVisitor { values }
     }
+}
 
-    /// Creates a new instance of a logger with the `trace` minimum level.
-    /// Enables tracing events to OTLP endpoint.
-    #[allow(dead_code)] // This is not ready for prime time yet!
-    pub fn new_with_telemetry(callback: ThreadsafeFunction<String>, endpoint: Option<String>) -> Self {
-        let javascript_cb = EventChannel::new(callback, EnvFilter::new("trace"), true);
+impl<'a> Visit for JsonVisitor<'a> {
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        match field.name() {
+            name if name.starts_with("r#") => {
+                self.values
+                    .insert(&name[2..], serde_json::Value::from(format!("{:?}", value)));
+            }
+            name => {
+                self.values
+                    .insert(name, serde_json::Value::from(format!("{:?}", value)));
+            }
+        };
+    }
 
-        global::set_text_map_propagator(TraceContextPropagator::new());
+    fn record_i64(&mut self, field: &Field, value: i64) {
+        self.values.insert(field.name(), serde_json::Value::from(value));
+    }
 
-        // A special parameter for Jaeger to set the service name in spans.
-        let resource = Resource::new(vec![KeyValue::new("service.name", "query-engine-node-api")]);
-        let config = Config::default().with_resource(resource);
+    fn record_u64(&mut self, field: &Field, value: u64) {
+        self.values.insert(field.name(), serde_json::Value::from(value));
+    }
 
-        let mut builder = opentelemetry_otlp::new_pipeline().tracing().with_trace_config(config);
-        let mut exporter = opentelemetry_otlp::new_exporter().tonic();
+    fn record_bool(&mut self, field: &Field, value: bool) {
+        self.values.insert(field.name(), serde_json::Value::from(value));
+    }
 
-        if let Some(endpoint) = endpoint {
-            exporter = exporter.with_endpoint(endpoint);
-        }
+    fn record_str(&mut self, field: &Field, value: &str) {
+        self.values.insert(field.name(), serde_json::Value::from(value));
+    }
+}
 
-        builder = builder.with_exporter(exporter);
+impl<'a> ToString for JsonVisitor<'a> {
+    fn to_string(&self) -> String {
+        serde_json::to_string(&self.values).unwrap()
+    }
+}
 
-        let tracer = builder.install_batch(opentelemetry::runtime::Tokio).unwrap();
+#[derive(Clone)]
+pub struct CallbackLayer {
+    callback: ThreadsafeFunction<String>,
+}
 
-        let telemetry_layer = tracing_opentelemetry::layer().with_tracer(tracer);
-        let registry = EventRegistry::new().with(telemetry_layer).with(javascript_cb);
-        let with_telemetry = WithTelemetry::new(registry);
+impl CallbackLayer {
+    pub fn new(callback: ThreadsafeFunction<String>) -> Self {
+        CallbackLayer { callback }
+    }
+}
 
-        let subscriber = Subscriber::WithTelemetry(with_telemetry);
+impl<S: Subscriber> Layer<S> for CallbackLayer {
```

Comment:
This is such a weird construct that I have no idea what is actually happening here. Can you elaborate what this code is doing?

Reply:
Sure, I will in a comment, long story short, instead of using a subscriber we are using our own specialized layer (layers are composable, subscribers are not). In the past we have our own subscriber and have different impementations depending if we want telemetry or not. With layers is not needed, you can compose them as a single happy friend. I will try to condensce this into a comment in the code, no worries.